### DISCUSSION TITLE
Parcel 2 Tree Shaking - (not) on by default?

### DISCUSSION THREAD

**Discussion by BrandonALittle** [parcel-bundler/parcel#5593] [Q&A] [answered]
Hi,

Running Parcel 2 like so: `PARCEL_BUNDLE_ANALYZER=1 parcel build src/index.tsx` with an `includeNodeModules: true` flag passed via a `targets` entry in `package.json` results in a very large bundle. Looking at the analyzer report, tree shaking does not seem to be happening, as many unused modules are being included. 

Could someone direct me to something that might help get tree shaking to work?

**Comment by mischnic** [comment, marked as answer]
1. Tree shaking is enabled by default
2. `includeNodeModules: true` is also the default setting

Can you give some code sample where unused code isn't removed?

**Reply by BrandonALittle** [reply]
![image](https://user-images.githubusercontent.com/19222963/103689257-e947c800-4f47-11eb-8b04-0b3dd541cbb7.png)
The vast majority of this is unused in our code base. Without the `targets` entry in `package.json`, none of the node modules get bundled with our code. Adding that resulted in the opposite extreme, with what you see above.

**Reply by BrandonALittle** [reply]
Just taking lodash as an example, we are using maybe 4 or 5 methods from lodash, but the bundle includes all of them.

**Reply by BrandonALittle** [reply]
Actually, we are using just three methods from lodash, imported via the following statements in different files:
`import { keyBy } from 'lodash';`
`import { debounce } from 'lodash';`
`import { isEqual } from 'lodash';`

**Reply by BrandonALittle** [reply]
This seems to provide an explanation, towards the bottom, as to why tree-shaking doesn't work with lodash (https://developers.google.com/web/fundamentals/performance/optimizing-javascript/tree-shaking), but as you can see in the report above, the same thing is happening with all of our node modules.

**Reply by mischnic** [reply]
For lodash, you should definitely use `lodash-es`. `import { keyBy, debounce, isEqual } from 'lodash-es';` should treeshake as expected.

> Without the targets entry in package.json, none of the node modules get bundled with our code

Then you probably have a `engines: {node` entry in package.json, which makes Parcel think you want to bundle e.g. a Node CLI app. Either remove that or set `context: "browser"` in the `targets` as well.

**Reply by BrandonALittle** [reply]
Indeed. Removing the `engines` entry seems to fix it. Thank you!
